fastqheat/__main__.py

```diff
@@ -6,18 +6,31 @@
 import ssl
 import subprocess
 from pathlib import Path
+from typing import Any
 
+import backoff
 import requests
 import urllib3
 
 from fastqheat.check import check_loaded_run, md5_checksum
+from fastqheat.config import get_settings, set_settings
 from fastqheat.metadata import ENAClient
 
 SRR_PATTERN = re.compile(r'^(SRR|ERR|DRR)\d+$')
 SRP_PATTERN = re.compile(r'^(((SR|ER|DR)[PAXS])|(SAM(N|EA|D))|PRJ(NA|EB|DB)|(GS[EM]))\d+$')
 USABLE_CPUS_COUNT = len(os.sched_getaffinity(0))
 
 
+@backoff.on_exception(
+    backoff.constant,
+    subprocess.CalledProcessError,
+    max_tries=lambda: get_settings().max_retries,
+)
+def _run_command(args: Any) -> None:
+    subprocess.run(args, check=True)
+
+
+@backoff.on_predicate(backoff.constant, max_tries=lambda: get_settings().max_retries)
```

Comment:
True, does seem verbose. Thank for the suggestion, although I think `backoff.on_predicate` will always use the default  value of `max_retries`, instead of one passed from the command line. Can you verify?

Reply:
I don't think it will use the default value, though it would be a good idea to test the code manually if you are not sure.

Here I found a brief explanation with examples:
https://python-course.eu/oop/class-instance-attributes.php

Reply:
Thanks for the link! I tested it, it seems to use the default value of 1.

Reply:
@dakq yeah, you are right, it is so because decorators are called at compile time, but setting up settings - at runtime. As there is no way to pass arguments to decorators dynamically, we could do something like this:

```python
MAX_TRIALS = 10  # use settings.max_trials instead


def backoff(mx_times):
    # just an example of backoff decorator
    def outer_wrapper(func):
        def inner_wrapper(*args, **kwargs):
            print(mx_times)
            return func(*args, **kwargs)
        return inner_wrapper
    return outer_wrapper


def _download_function_ftp(foo, bar):
    # real download ftp function
    return foo + bar


def _download_function_fasterq_dump(foo, bar):
    # real download fasterq_dump function
    return foo + bar


def download_function_ftp(foo=1, bar=2):
    # wrap the backoff decorator to be able to use max_trials that is evaluated at runtime
    return backoff(mx_times=MAX_TRIALS)(_download_function_ftp)(foo=foo, bar=bar)


def download_function_fasterq_dump(foo=1, bar=2):
    # wrap the backoff decorator to be able to use max_trials that is evaluated at runtime
    return backoff(mx_times=MAX_TRIALS)(_download_function_fasterq_dump)(foo=foo, bar=bar)


print(download_function_ftp(1, 1))
print(download_function_fasterq_dump(2, 2))
```

I am not 100% sure that it is the ideal solution, but if we don't have a better one for now, we can do it like this

Reply:
Nice if a little complicated solution. To clarify, do you think it's better than passing a callable to `backoff`'s `max_tries` argument, like in my original (verbose) solution? `backoff` explicitly allows users use callables to provide arguments to avoid this kind of problem.

Reply:
@dakq in this case let's leave it as you have already made it: 

```
@backoff.on_predicate(backoff.constant, max_tries=lambda: get_settings().max_retries)
```